PhysicsTools/PatAlgos/plugins/PATLostTracks.cc

```diff
@@ -255,16 +273,100 @@ void pat::PATLostTracks::addPackedCandidate(std::vector<pat::PackedCandidate>& c
+      }
+    }
+  } else if (!useLegacySetup_ && trk->pt() > 0.5) {
+    if (trk->hitPattern().numberOfValidPixelHits() > 0) {
+      cands.back().setTrackProperties(
+          *trk, covariancePackingSchemas_[2], covarianceVersion_);  // low quality with pixels
+    } else {
+      cands.back().setTrackProperties(
+          *trk, covariancePackingSchemas_[3], covarianceVersion_);  // low quality without pixels
+    }
+  }
+  cands.back().setAssociationQuality(pvAssocQuality);
+}
+
+std::pair<int, pat::PackedCandidate::PVAssociationQuality> pat::PATLostTracks::associateTrkToVtx(
+    const reco::VertexCollection& vertices, const reco::TrackRef& trk) const {
+  //For legacy setup check only if the track is used in fit of the PV, i.e. vertices[0],
+  //and associate quality if weight > 0.5. Otherwise return invalid vertex index (-1)
+  //and default quality flag (NotReconstructedPrimary = 0)
+  if (useLegacySetup_) {
+    float w = vertices[0].trackWeight(trk);
+    if (w > 0.5) {
+      return std::pair<int, pat::PackedCandidate::PVAssociationQuality>(0, pat::PackedCandidate::UsedInFitTight);
+    } else if (w > 0.) {
+      return std::pair<int, pat::PackedCandidate::PVAssociationQuality>(0,
+                                                                        pat::PackedCandidate::NotReconstructedPrimary);
+    } else {
+      return std::pair<int, pat::PackedCandidate::PVAssociationQuality>(-1,
+                                                                        pat::PackedCandidate::NotReconstructedPrimary);
+    }
+  }
+
+  //Inspired by CommonTools/RecoAlgos/interface/PrimaryVertexAssignment.h
+  //but without specific association for secondaries in jets and option to use timing
+
+  int iVtxMaxWeight = -1;
+  int iVtxMinDzDist = -1;
+  size_t idx = 0;
+  float maxWeight = 0;
+  double minDz = std::numeric_limits<double>::max();
+  double minDzSig = std::numeric_limits<double>::max();
+  for (auto const& vtx : vertices) {
+    float w = vtx.trackWeight(trk);
+    double dz = std::abs(trk->dz(vtx.position()));
+    double dzSig = dz / trk->dzError();
+    if (w > maxWeight) {
+      maxWeight = w;
+      iVtxMaxWeight = idx;
+    }
+    if (dzSig < minDzSig) {
+      minDzSig = dzSig;
+      minDz = dz;
+      iVtxMinDzDist = idx;
+    }
+    idx++;
+  }
+  // vertex in which fit the track was used
+  if (iVtxMaxWeight >= 0) {
+    if (maxWeight > 0.5) {
+      return std::pair<int, pat::PackedCandidate::PVAssociationQuality>(iVtxMaxWeight,
+                                                                        pat::PackedCandidate::UsedInFitTight);
+    } else {
+      return std::pair<int, pat::PackedCandidate::PVAssociationQuality>(iVtxMaxWeight,
+                                                                        pat::PackedCandidate::UsedInFitLoose);
+    }
+  }
+  // vertex "closest in Z" with tight cuts (targetting primary particles)
+  const double add_cov = vertices[iVtxMinDzDist].covariance(2, 2);
```

Comment:
My previous suggestion was the following (otherwise the index `iVtxMinDzDist` is not ensured to be positive):
```suggestion
  if (minDz < maxDzForPrimaryAssignment_}
    const double add_cov = vertices[iVtxMinDzDist].covariance(2, 2);
    const double dzErr = sqrt(trk->dzError() * trk->dzError() + add_cov);
    if (minDz / dzErr < maxDzSigForPrimaryAssignment_ && trk->dzError() < maxDzErrorForPrimaryAssignment_) {
      return std::pair<int, pat::PackedCandidate::PVAssociationQuality>(iVtxMinDzDist,
                                                                      pat::PackedCandidate::CompatibilityDz);
    }
  }
```

Reply:
OK, I will change it like this. But, I realized that by construction `iVtxMinDzDist >= 0` if there is at least one vertex as dz significance it should be smaller than `std::numeric_limits<double>::max()`